src/security/Handshake.cc

```diff
@@ -631,33 +634,37 @@ Security::HandshakeParser::parseSupportedVersionsExtension(const SBuf &extension
 void
 Security::HandshakeParser::skipMessage(const char *description)
 {
-    // tkMessages/fragments can only contain messages of the same ContentType.
+    // skip at least one byte
+    if (tkMessages.atEnd())
+        throw Parser::InsufficientInput();
+    // tkMessages can only contain messages of the same ContentType.
     // To skip a message, we can and should skip everything we have [left]. If
-    // we have partial messages, debugging will mislead about their boundaries.
+    // we buffered a partial message, we will need to read/skip multiple times.
     tkMessages.skip(tkMessages.leftovers().length(), description);
 }
 
 bool
 Security::HandshakeParser::parseHello(const SBuf &data)
 {
-    try {
-        if (!expectingModernRecords.configured())
-            expectingModernRecords.configure(!isSslv2Record(data));
+    if (!expectingModernRecords.configured())
+        expectingModernRecords.configure(!isSslv2Record(data));
+
+    // data contains everything read so far, but we may read more later
+    tkRecords.reinput(data, true);
+    tkRecords.rollback();
+    while (!done) {
+        try {
+            if (tkRecords.atEnd())
+                return false;
 
-        // data contains everything read so far, but we may read more later
-        tkRecords.reinput(data, true);
-        tkRecords.rollback();
-        while (!done)
             parseRecord();
-        debugs(83, 7, "success; got: " << done);
-        // we are done; tkRecords may have leftovers we are not interested in
-        return true;
-    }
-    catch (const Parser::BinaryTokenizer::InsufficientInput &) {
-        debugs(83, 5, "need more data");
-        return false;
```

Comment:
This "return false" for InsufficientInput case was correct, and the caller relies on it AFAICT: We have to keep it or move InsufficientInput processing to the caller. The added atEnd() check above is another red flag. Overall, I am not sure that this method refactoring improved things, but perhaps I am missing some essential changes in this diff noise.

Reply:
Unfortunately, these changes failed one of my per-byte test - the tkRecords.atEnd() check does not work when tkRecords have some unparsed bytes insufficient for the next record. We need to know somehow who threw InsufficientInput() in the catch() below - and, depending on that, either return false (insufficient tkRecords) or continue to the next record (insufficient tkMessages). I tried to add some synced() method  (whether all parsed bytes were comitted) into BinaryTokenizer and calculate this state as (!synced() || atEnd()) for tkMessages. But it looks a bit complicated/unreliable - we should not calculate the state again with the exception at hand. Unfortunately, InsufficientInput() does not have any attribute about its initiator.

Reply:
> either return false (insufficient tkRecords) or continue to the next record (insufficient tkMessages). ... Unfortunately, InsufficientInput() does not have any attribute about its initiator.

To treat InsufficientInput emitted by message layer differently, catch InsufficientInput emitted by the that layer at that layer (i.e. when we parse messages). InsufficientInput caught while making parseRecord() calls will be specific to the records layer. AFAICT, InsufficientInput emitted while parsing messages should be "caught, reported, and ignored" -- the control will then bubble up to the next parseRecord() loop iteration.

Alternatively, we could parse all records we have read _before_ parsing the messages in the accumulated fragments. If message layer always sees all available same-type records, then InsufficientInput thrown by any layer will have the same meaning -- need to read more bytes (record bytes _and_ message bytes). However, it is not clear to me whether such refactoring will make the code better or worse overall. Please try to fix the current (i.e. "parse one record fragment at a time") code _before_ we attempt such a significant refactoring.

Reply:
>  InsufficientInput emitted by the that layer at that layer 

Ok. I hesitated doing this because recently we [rejected](https://github.com/measurement-factory/squid/pull/240/files/095601c5bcda15a41efd3685496aabde26a281b0#r1430223924) tkMessages try/catch for a reason.